src/lem/circuit.rs

```diff
@@ -0,0 +1,889 @@
+    store::Store,
+    MetaPtr, LEM, LEMOP,
+};
+
+/// Holds a counter per path. We want to use this to count the number of slots
+/// that have already been used on each LEM path.
+#[derive(Default)]
+struct PathTicker(HashMap<Path, usize>);
+
+impl PathTicker {
+    /// Increments the counter of a path. If the path wasn't tracked, returns
+    /// `0` and starts tracking it
+    pub(crate) fn tick(&mut self, path: Path) -> usize {
+        let next = self.0.get(&path).unwrap_or(&0).to_owned();
+        self.0.insert(path, next + 1);
+        next
+    }
+
+    /// Starts tracking a new path with a counter from another. If the reference
+    /// path wasn't being tracked, the new one won't be either, such that calling
+    /// `next` will return `0`.
+    pub(crate) fn cont(&mut self, new: Path, from: &Path) {
+        match self.0.get(from) {
+            Some(i) => {
+                self.0.insert(new, *i);
+            }
+            None => (),
+        }
+    }
+}
+
+/// Keeps track of previously used slots indices for each possible LEM path,
+/// being capable of informing the next slot index to be used
+#[derive(Default)]
+struct SlotsCounter {
+    hash2: PathTicker,
+    hash3: PathTicker,
+    hash4: PathTicker,
+}
+
+impl SlotsCounter {
+    #[inline]
+    pub(crate) fn next_hash2(&mut self, path: Path) -> usize {
+        self.hash2.tick(path)
+    }
+
+    #[inline]
+    pub(crate) fn next_hash3(&mut self, path: Path) -> usize {
+        self.hash3.tick(path)
+    }
+
+    #[inline]
+    pub(crate) fn next_hash4(&mut self, path: Path) -> usize {
+        self.hash4.tick(path)
+    }
+
+    pub(crate) fn cont(&mut self, new: Path, from: &Path) {
+        self.hash2.cont(new.clone(), from);
+        self.hash3.cont(new.clone(), from);
+        self.hash4.cont(new, from);
+    }
+}
+
+/// Contains a `slots_map` that maps `LEMOP`s to their slots. This map is not
+/// expected to be injective, as two or more `LEMOP`s can be mapped to the same
+/// slot. The `slots` attribute is an set of all slots present on `slots_map`.
+pub struct SlotsInfo {
+    slots_map: IndexMap<LEMOP, (usize, SlotType)>,
+    slots: IndexSet<(usize, SlotType)>,
+}
+
+impl SlotsInfo {
+    #[inline]
+    pub fn get_slot(&self, op: &LEMOP) -> Result<&(usize, SlotType)> {
+        self.slots_map
+            .get(op)
+            .ok_or_else(|| anyhow!("Slot not found for LEMOP {:?}", op))
+    }
+}
+
+impl LEMOP {
+    /// STEP 1: compute the slot mapping on a first (and unique) traversal
+    ///
+    /// While traversing parallel paths, we need to reuse compatible slots. For
+    /// example, on a `MatchTag` with two arms, such that each arm is a `Hash2`,
+    /// we want those hashes to occupy the same slot. We do this by using a
+    /// `SlotsCounter`, which can return the next index for a slot category
+    /// (`Hash2`, `Hash3` etc).
+    ///
+    /// Further, when we:
+    /// * Construct the same pointer twice
+    /// * Destruct the same pointer twice
```

Comment:
I think you mean 'deconstruct' rather than 'destruct'. When you actually mean hashing, as opposed to any possible slot usage, I think 'destructure' is the better word — leaning into our use of hash-consing.